### ISSUE TITLE
Projected volume mounts do not respect defaultMode

### ISSUE THREAD

**Issue by howardjohn** [kubernetes/kubernetes#89153]
**What happened**: I used a projected volume mount with `defaultMode: 0777`, and the mode was not changed.

**What you expected to happen**: The goal of this was to allow a container that has `runAsGroup` set to be able to read this.

**How to reproduce it (as minimally and precisely as possible)**:

Apply pod:
```yaml
cat <<EOF | kaf -
apiVersion: v1
kind: Pod
metadata:
  name: alpine-no-context
spec:
  containers:
  - args:
    - sleep
    - "10000000000000000"
    image: alpine
    name: alpine
    securityContext:
      runAsGroup: 1234
      runAsUser: 1234
    volumeMounts:
    - mountPath: /var/run/secrets/tokens
      name: token
  volumes:
  - name: token
    projected:
      defaultMode: 0777
      sources:
      - serviceAccountToken:
          audience: foo
          expirationSeconds: 43200
          path: my-token
EOF
```

Exec onto the pod:

```
/ $ cd /run/secrets/tokens/..data
/run/secrets/tokens/..2020_03_16_17_04_24.856732978 $ ls -la
total 4
drwxr-xr-x    2 root     root            60 Mar 16 17:04 .
drwxrwxrwt    3 root     root           100 Mar 16 17:04 ..
-rw-------    1 root     root           882 Mar 16 17:04 my-token
/run/secrets/tokens/..2020_03_16_17_04_24.856732978 $ cat my-token
cat: can't open 'my-token': Permission denied
/run/secrets/tokens/..2020_03_16_17_04_24.856732978 $ whoami
whoami: unknown uid 1234
/run/secrets/tokens/..2020_03_16_17_04_24.856732978 $
```

As you can see the token is mounted with `0400`, the default, rather than `0777`. As a result, I cannot read the token.

This can be fixed by adding `securityContext.fsGroup=1234`, but this applies to ALL volumes, even in other containers. This causes issues like https://github.com/istio/istio/issues/22025, where an unrelated container is trying to mount a volume.

**Comment by neolit123** [comment]
please share the affected kubernetes versions.
/sig storage

**Comment by howardjohn** [comment]
I have tested GKE 1.15 and kind 1.17, both the same behavior

**Comment by rlenglet** [comment]
Is this a duplicate of https://github.com/kubernetes/kubernetes/issues/57923?

**Comment by mgxian** [comment]
Can someone fix this?

**Comment by alanbrent** [comment]
To add to the affected versions question: We're experiencing this in our EKS clusters, currently running version `v1.14.9-eks-f459c0`.

>Is this a duplicate of #57923?

AFAICT that issue explicitly refers to using `pod.spec.securityContext.runAsUser`, which we are **not** doing. Instead, the container image sets the execution user. While the effect is roughly the same, the relevant Pod configuration is different so I probably wouldn't call it a duplicate.

**Comment by adracus** [comment]
Seems to be fixed in K8s >= 1.19: https://github.com/kubernetes/kubernetes/commit/3db7275b549559696c42c0b5f51c9a2397e9571d
